File: fiap/python/projeto_agenda.py
agenda = []  # Criando uma lista vazia.

# DEFINIÇÃO DAS FUNÇÕES

# Descrição: Este procedimento cria um novo contato na agenda.
# Nome: novo()
# Tipo: procedimento
def novo():
    global agenda  # Definindo variável como Global.
    nome = p_nome()
    celular = input("Celular....: ")
    email   = input("E-mail.....: ")
    agenda.append([nome, celular, email])  # Adicionando os dados na agenda
    print("-----------------------------"
          "Registro gravado com Sucesso!!!"
          "-------------------------------")

# Descrição: Este procedimento lê um nome.
# Nome: p_nome()
# Tipo: procedimento
def p_nome():
    return(input("Nome.......: "))

# Descrição: Este procedimento lista um registro
# Nome: listar_dados(nome, celular, email)
# Tipo: procedimento
def listar_dados(nome, celular, email):
    print("Nome: %sCelular: %sEmail: %s" % (nome, celular, email))
    print("----------------------------------")

# Descrição: Este procedimento lista todos os registros da matriz.
# Nome: listar()
# Tipo: procedimento
def listar():  # Função para mostrar lista de contatos.
    print("CONTATOS DA AGENDA ###############")
    for e in agenda:
        listar_dados(e[0], e[1], e[2])
    print("FIM DA AGENDA ####################")

# Descrição: Esta função pesquisa um contato pelo nome.
# Nome: pesquisa(nome): int
# Tipo: função
def pesquisa(nome):  # Função para pesquisar contatos.
    name = nome.lower()
    for d, e in enumerate(agenda):  # percorre toda a matriz.
        if e[0].lower() == name:  # procura o nome desejado
            return d  # retorna o índice do nome encontrado
    return None  # retorna vazio se não encontrar

# Descrição: Este procedimento exibe o registro ou mensagem de insucesso
# Nome: pesquisar():
# Tipo: procedimento
def pesquisar():
    # pesquisa o nome
    p = pesquisa(p_nome())  # Entrada de dados.
    if p != None:
        print("Registro encontrado!")
        # atualiza as variáveis se encontrou
        nome = agenda[p][0]
        celular = agenda[p][1]
        email = agenda[p][2]
        # mostra o registro
        listar_dados(nome, celular, email)
    else:
        # exibe a mensagem de insucesso na procura do registro
        print("Nome não encontrado!!!")

# Descrição: Este procedimento apaga um contato
# Nome: apagar():
# Tipo: procedimento
def apagar():
    global agenda
    nome = p_nome()
    # retorna o índice do nome ou vazio
    p = pesquisa(nome)
    if p != None:  # Se encontrou o contato
        del agenda[p]  # exclui o contato
        print("-----------------------------"
              "Registro APAGADO com Sucesso!!!"
              "-------------------------------")
    else:
        # não encontrou o registro para excluir
        print("Nome não encontrado.")

# Descrição: Este procedimento edita um contato
# Nome: editar():
# Tipo: procedimento
def editar():
    p = pesquisa(p_nome())  # Entrada de dados.
    # Se encontrou o registro
    if p != None:
        # mostra o nome e pede a edição dos demais
        nome = agenda[p][0]
        print("Nome........: ", nome)
        celular = input("Celular....:")
        email = input("E-mail.....:")
        agenda[p] = [nome, celular, email]  # Armazenando os novos dados.
        print("-----------------------------"
              "Registro EDITADO com Sucesso!!!"
              "-------------------------------")
    else:
        print("Nome não encontrado.")  # Executa caso a condição seja falsa.

# Descrição: Esta função valida SE O ITEM DIGITADO FOI VALIDO
# Nome: validar(pergunta, inicia, fim): int
# Tipo: função
def validar(pergunta, inicio, fim):  # Função para validar números inteiros.
    while True:  # Criando um loop infinito.
        try:  # Criando um acordo/condição.
            valor = int(input(pergunta))  # Entrada de dados.
            if inicio <= valor <= fim:  # Determinando uma condição.
                return (valor)  # Executa caso for verdadeira.
            else:
                return (0)
        except ValueError:  # Executa caso for falsa.
            print("Valor inválido, favor digitar entre %d e %d" % (inicio, fim))

# Descrição: Esta função retorna o item do menu ou 0 para invalido
# Nome: menu(pergunta, inicia, fim): int
# Tipo: função
def menu():  # Função que exibe o menu de opções.
    print("""
   1 - Adicionar novo contato
   2 - Editar um contato
   3 - Pesquisar contato
   4 - Lista de contatos
   5 - Apagar um contato
   6 - Sair
""")
    return validar("Escolha uma opção: ", 1, 6)

while True:  # Criando um loop infinito.
    opcao = menu()
    if opcao == 0:
        print("Opcao Inválida!")
    elif opcao == 6:
        break
    elif opcao == 1:
        novo()
    elif opcao == 2:
        editar()
    elif opcao == 3:
        pesquisar()
    elif opcao == 4:
        listar()
    elif opcao == 5:
        apagar()

File: fiap/java/Projeto_agenda.java
package fiap.java;
import java.util.Scanner;

public class Projeto_agenda {

    // SUBALGORITMOS
    // Variáveis e objetos públicos
    public static Scanner entrada = new Scanner(System.in);
    public static String agenda[][] = new String[10][3];

    /*
     * Descrição.....: Este procedimento limpa a matriz
     * Nome..........: limpar_matriz(String matriz)
     * Tipo..........: void
     */
    public static void limparMatriz(String mm[][]) {
        // Insere vazio em todas as células da matriz
        for (int l = 0; l <10; l++) {
            for (int c = 0; c < 3; c++) {
                mm[l][c] = "";
            }
        }
    }

    /*
     * Descrição....: Este procedimento cadastra um novo contato
     * Nome.........: novo(String matriz, int linha)
     * Tipo.........: void
     */
    public static void novo(String mm[][], int l) {
        // Pede ao usuário a edição dos campos da linha passada por paramêtro
        System.out.println("---PREENCHA O NOVO CONTATO: ");
        System.out.println("Nome...: ");
        mm[l][0] = entrada.next();
        System.out.print("Celular...: ");
        mm[l][1] = entrada.next();
        System.out.print("E-mail...:");
        mm[l][2] = entrada.next();
    }

    /*
     * Descrição....: Este procedimento edita um contato
     * Nome.........: editarContato(String matriz, int linha)
     * Tipo.........: void
     */
    public static void editarContato(String mm[][], int l) {
        // Permite ao usuário editar o contato encontrado
        // Não permite que o nome seja editado por ser a chave de pesquisa
        System.out.println("---EDITE O CONTATO: ");
        System.out.print("Nome...: "+ mm[l][0]);
        System.out.print("Celular...: ");
        mm[l][1] = entrada.next();
        System.out.print("E-mil...: ");
        mm[l][2] = entrada.next();
    }

    /*
     * Descrição....: Esta função retorna a próxima linha em branco da matriz
     * Nome.........: linhaProximoContato(String matriz, int linha)
     * Tipo.........: int
     */
    public static int linhaProximoContato(String mm[][]) {
        for (int l = 0; l < 10; l++) {
            if (mm[l][0].equals("")) {
                // Caso tenha encontrado, retorne o número da linha
                // Próxima linha vazia
                return l;
            }
        }
        // -1 representa a matriz estar cheia
        return -1;
    }

    /*
     * Descrição....: Este procedimento exibe um contato
     * Nome.........: exibirContato(String matriz, int linha)
     * Tipo.........: void
     */
    public static void exibirContato(String mm[][], int linha) {
        // Exibe o registro da linha passada por parâmetro
        System.out.println("Nome...: "+ mm[linha][0]);
        System.out.println("Celular...: "+ mm[linha][1]);
        System.out.println("E-mail...: "+ mm[linha][2]);
    }

    /*
     * Descrição....: Este procedimento lista todos os contatos cadastrados
     * Nome.........: listarAgenda(String matriz)
     * Tipo.........: void
     */
    public static void listarAgenda(String mm[][]) {
        System.out.println("--- CONTATOS DA AGENDA: ");
        for (int l = 0; l < 10; l++) {
            // Enquanto tiver registro, exibe-o
            if (mm[l][0] != "") {
                exibirContato(mm, l);
                System.out.println("----------");
            }
        }
        System.out.println("--- FIM DA AGENDA: ");
    }

    /*
     * Descrição.....: Esta função pesquisa o contato e retorna a linha
     * Nome..........: pesquisarContato(String matriz, string nome)
     * Tipo..........: int (retorna -1 se não encontrou)
     */
    public static int pesquisarContato(String mm[][], String n) {
        // Caso encontre o contato, retorna a linha onde ele está
        for (int l = 0; l < 10; l++) {
            if (mm[l][0].equals(n)) {
                return l;
            }
        }
        return -1;
    }

    /*
	 * Descrição....: Este procedimento exclui a linha passada por parâmetro
	 * Nome.........: ExcluiLinha(String matriz, int linha)
	 * Tipo.........: void
	*/
	public static void excluiLinha(String mm[][], int l) {
		// exclui o contato a partir da linha passada por parâmetro
		mm[l][0] = "";
		mm[l][1] = "";
		mm[l][2] = "";
		System.out.println("Contato Excluído");
	}

    /*
	 * Descrição....: Este procedimento pesquisa e exclui um contato
	 * Nome.........: apagarContato(String matriz, String nome)
	 * Tipo.........: void
	*/
	public static void apagarContato(String mm[][], String n) {
		// Inicia a variável que achou como false
		boolean achou = false;
		// Alimenta a linha com o valor da pesquisa (-1 não encontrou)
		int linha = pesquisarContato(mm, n);
		String opcao;
		if (linha != -1) {
			// Exibe o contato a partir da linha
			exibirContato(mm,linha);
			// Confirma com o usuário se ele quer excluir ou nao o contato
			System.out.println("Confirma a exclusão do contato? [S]im ou [N]ão?");
			opcao = entrada.next();
			if (opcao.equals("s") || opcao.equals("S")) {
				// Se a escolha for Sim, exclui o contato
				excluiLinha(mm, linha);
			} else {
				// Cancela a exclusão
				System.out.println("Exclusão cancelada!");
			}
		} else {
			// Contato não encontrado
			System.out.println("Contato não encontrado!");
		}

	}

    public static void exibeMenu() {
		System.out.println("********* M E N U ********");
		System.out.println("1 - Adicionar novo contato");
		System.out.println("2 - Editar contato");
		System.out.println("3 - Pesquisar contato");
		System.out.println("4 - Lista de contatos");
		System.out.println("5 - Apagar um contato");
		System.out.println("6 - Sair");
	}

	public static void main(String[] args) {
		int opcao, linha;
		String nome;

		// ao iniciar a aplicação, sempre é bom limparmos a matriz para que a "sujeira"
		// do Buffer não influencie nos resultados
		limparMatriz(agenda);

		do {
			// Exibição do menu
			exibeMenu();

			// Colhe a opção escolhida pelo usuário
			System.out.print("Escolha uma opção:");
			opcao = entrada.nextInt();
			System.out.println();

			// Seleciona a opção escolhida
			switch(opcao) {
				// Caso a escolha seja "adicionar novo contato"
				case 1:
					novo(agenda, linhaProximoContato(agenda));
					break;

				// Caso a escolha seja "Editar contato"
				case 2:
					System.out.println("------------ EDITANDO (PESQUISE O CONTATO): ");
					System.out.print("Digite o nome.........:");
					nome = entrada.next();
					linha = pesquisarContato(agenda, nome);
					if (linha == -1) {
						// informa se não encontrou o contato
						System.out.print("Contato não Cadastrado!");
					} else {
						// se encontrou, exibe o contato e o edite
						exibirContato(agenda, linha);
						editarContato(agenda, linha);
					}
                    break;

				// Efetua a pesquisa do contato
				case 3:
					// pede o nome
					System.out.println("------------ PESQUISE O CONTATO: ");
					System.out.print("Digite o nome.........:");
					nome = entrada.next();
					// Retorna a linha da pesquisa
				    linha = pesquisarContato(agenda, nome);
					if (linha == -1) {
						// Se o contato não existe
						System.out.print("Contato não Cadastrado!");
					} else {
						// Se o contato existe, exibí-lo
						exibirContato(agenda, linha);
					}
					break;

				// Lista todos os Contatos da agenda
				case 4:
					listarAgenda(agenda);
				    break;

				// Exclui um registro digitado pelo usuário
				case 5:
					System.out.println("------------ EXCLUINDO (PESQUISE O CONTATO): ");
					System.out.print("Digite o nome.........:");
					nome = entrada.next();
					apagarContato(agenda, nome);
					break;

				case 6:
					System.out.print("OBRIGADO POR UTILIZAR A NOSSA AGENDA :)");

			}
			System.out.println();
		}while(opcao != 6);
	}
}
